docs/developers/developing_iree/sanitizers.md

```diff
@@ -23,39 +23,66 @@ Notes:
   documentation](https://clang.llvm.org/docs/AddressSanitizer.html#memory-leak-detection)
   on leak detection. It is only enabled by default on some platforms.
 
-## Enabling the sanitizers
+## Support status and how to enable each sanitizer
 
-In the CMake build system of IREE, at least on Linux and Android, enabling these
-sanitizers is a simple matter of passing one of these options to the initial
-CMake command:
+### ASan (AddressSanitizer)
+
+Enabling ASan in the IREE build is a simple matter of setting the
+`IREE_ENABLE_ASAN` CMake option:
 
 ```
--DIREE_ENABLE_ASAN=ON
+cmake -DIREE_ENABLE_ASAN=ON ...
 ```
 
-or
+### TSan (ThreadSanitizer)
+
+Enabling TSan in the IREE build is, in itself, a simple matter of setting the
+`IREE_ENABLE_TSAN` CMake option:
 
 ```
--DIREE_ENABLE_MSAN=ON
+cmake -DIREE_ENABLE_TSAN=ON ...
 ```
 
-or
+That enables TSan in all the C/C++ targets, which includes the IREE runtime,
+the IREE compiler, and all the C/C++ unit tests.
+
+The resuting TSan-instrumented IREE runtime is only able to load IREE modules
+that have also been built with TSan instrumentation, as enabled by passing
+`iree-llvm-sanitize=thread` to the IREE compiler. That in turn also requires
+passing `-iree-llvm-link-embedded=false` to the IREE compiler.
+
+Many of IREE's own tests involve such an IREE module. Passing
+`IREE_ENABLE_TSAN=ON` does not in itself change the flags passed to the IREE
+compiler when building such test modules.
+
+When making a IREE build that involves IREE modules (such as tests), the
+`IREE_BYTECODE_MODULE_ENABLE_TSAN` and
+`IREE_BYTECODE_MODULE_FORCE_SYSTEM_DYLIB_LINKER` CMake options must also be set,
+as in the following command:
 
 ```
--DIREE_ENABLE_TSAN=ON
+cmake -DIREE_ENABLE_TSAN=ON \
+  -DIREE_BYTECODE_MODULE_ENABLE_TSAN=ON \
+  -DIREE_BYTECODE_MODULE_FORCE_SYSTEM_DYLIB_LINKER=ON \
+  ...
 ```
 
-These sanitizers will be most helpful on builds with debug info, so consider
-using
+### MSan (MemorySanitizer)
+
+In theory that should be a simple matter of
 
 ```
--DCMAKE_BUILD_TYPE=RelWithDebInfo
+-DIREE_ENABLE_MSAN=ON
 ```
 
-instead of just `Release`. It's also fine to use sanitizers on `Debug` builds,
-of course --- if the issue that you're tracking down reproduces at all in a
-debug build! Sanitizers are often used to track down subtle issues that may only
-manifest themselves in certain build configurations.
+However, that requires making and using a custom
+build of libc++ with MSan as explained
+[here](https://github.com/google/sanitizers/wiki/MemorySanitizerLibcxxHowTo).
+
+As of April 2022, all of IREE's tests succeeded with MSan on Linux/x86-64,
+provided that the `vulkan` driver was disabled (due to lack of MSan
+instrumentation in the NVIDIA Vulkan driver).
```

Comment:
Curious if this applies for other drivers (e.g. CUDA). Might help to just say LLVM/CPU.

Reply:
In my understanding MSan is a "rebuild the world" sanitizer, more so even than other sanitizers. The [MemorySanitizerLibcxxHowTo](https://github.com/google/sanitizers/wiki/MemorySanitizerLibcxxHowTo) link being added in the docs here makes the point: to use MSan, you need everything rebuilt with MSan down to the standard C++ library. As GPU drivers are ultimately CPU code (that happens to talk to a GPU), they too would need to be rebuilt from source together with their own dependencies. That was my understanding of why MSan sounded pretty much antagonistic to any GPU backend. Even if the entire driver is open source, it would still take a lot of effort to rebuild it all with MSan. When the driver is not open source, in my understanding, it's a nonstarter.

Reply:
Hmm... alright, though GPU drivers aren't typically statically linked into the application. Some of the sanitizers have carve-outs for dynamic libraries as those are out of the control of the application being inspected.

Reply:
I actually tried it, it actually fails as mentioned - due to lack of MSan instrumentation in NVIDIA Vulkan driver.

I don't believe that it matters that GPU drivers aren't statically linked. The necessary MSan instrumentation is a property of what the code inherently does (bookkeeping to track uninitialized memory), not of how it gets linked.